### PR TITLE
feat(service): Make the service check the http code response and log …

### PR DESCRIPTION
**Pull request by camilamacedo86**
## Motivation
https://github.com/aerogear/mobile-security-service-operator/pull/68

## What
* Make the service validate/check the HTTP code status response
* Promote better logs for the failure situations

## Verification Steps
- It can be checked with this PR and `docker.io/cmacedo/mobile-security-service-operator:FIX-SERVICE` 
- Install the operator `make create-all`
- Run `oc new-project mobile-security-service-apps`
- Then in the valid app namespace try to create an app `make create-app`
- Then in the valid app namespace try to remove the app `make delete-app`

IMPORTANT: It cannot be checked if all works. So, if we the issue be solved in the master branch, https://github.com/aerogear/mobile-security-service-operator/pull/76, and its be rebase we will not able to check what was done here.

Screenshot 2019-05-11 at 00 37 02

## Checklist:

- [X] Code has been tested locally by PR requester
- [ ] Changes have been successfully verified by another team member 

## Progress

- [x] Finished task
- [ ] TODO

### CODE REVIEW

File: pkg/service/client.go [hunk 1]
```diff
@@ -10,13 +10,13 @@ import (
 
 //DeleteAppFromServiceByRestAPI delete the app object in the service
 func DeleteAppFromServiceByRestAPI(serviceAPI string, id string, reqLogger logr.Logger ) error {
-	reqLogger.Info( "Calling REST API to DELETE app", "serviceAPI", serviceAPI, "App.id", id )
+	reqLogger.Info( "Calling REST Service to DELETE app", "serviceAPI", serviceAPI, "App.id", id )
```
Comment (damienomurchu) [inline pkg/service/client.go]:
```suggestion
	reqLogger.Info( "Calling Service to DELETE app", "serviceAPI", serviceAPI, "App.id", id )
```

Is there a need to specify `REST`? This seems an implementation detail of the client, and if we swap out the current REST implementation in the client for something for something else, like if we changed to using GraphQL it seems we would need to change a lot of logging comments.

**Comment by damienomurchu** [comment]
Changes look good aside from one comment left re use of `REST` in log messages. Haven't verified though.

File: pkg/service/client.go (hunk 1 above)
Comment (camilamacedo86) [reply pkg/service/client.go]: The logs are important for we are able to troubleshoot it. 
As you can see the changes performed here is more to keep all texts equals and check the HTTP code status for we have sure if it worked or not as throw the errors.

Comment (damienomurchu) [reply pkg/service/client.go]: @camilamacedo86 I'm not sure you understood my point - can you re-read above comment

File: pkg/service/client.go [hunk 2]
```diff
@@ -77,44 +78,42 @@ func GetAppFromServiceByRestApi(serviceAPI string, appId string, reqLogger logr.
...
 	}
 
 	//Do the request
 	client := &http.Client{}
 	response, err := client.Do(req)
 	defer response.Body.Close()
-
 	if err != nil {
-		reqLogger.Error(err, "Error when try to do the request", "HTTPMethod", http.MethodGet, "Request", req, "url", url, "error", err)
+		reqLogger.Error(err, "Unable to do exec GET request", "HTTPMethod", http.MethodGet, "url", url, "response.StatusCode", response.StatusCode)
```
Comment (laurafitzgerald) [inline pkg/service/client.go]: Suggest -> `Unable to execute GET request`

File: pkg/service/client.go [hunk 3]
```diff
@@ -125,7 +124,7 @@ func GetAppFromServiceByRestApi(serviceAPI string, appId string, reqLogger logr.
 
 //deleteAppFromServiceByRestAPI delete the app object in the service
 func UpdateAppNameByRestAPI(serviceAPI string, app models.App, reqLogger logr.Logger) error {
-	reqLogger.Info( "Calling REST API to PATCH app", "serviceAPI", serviceAPI, "App", app )
+	reqLogger.Info( "Calling Rest Service to PATCH app", "serviceAPI", serviceAPI, "App", app )
```
Comment (laurafitzgerald) [inline pkg/service/client.go]: By PATCH here do you mean `update`? Could we use update or POST rather than PATCH?

File: pkg/service/client.go [hunk 4]
```diff
@@ -139,19 +138,19 @@ func UpdateAppNameByRestAPI(serviceAPI string, app models.App, reqLogger logr.Lo
 	req, err := http.NewRequest(http.MethodPatch, url ,strings.NewReader(string(appJSON)))
 	req.Header.Set("Content-Type", "application/json")
 	if err != nil {
-		reqLogger.Error(err, "Error when try to create the request", "HTTPMethod", http.MethodPatch, "Request", req, "url", url, "error", err)
+		reqLogger.Error(err, "Unable to create PATH request to update app name", "HTTPMethod", http.MethodPatch, "url", url)
```
Comment (laurafitzgerald) [inline pkg/service/client.go]: I think `PATH` is a typo here but also see above comment about the use of `PATCH`

File: pkg/service/client.go [hunk 5]
```diff
@@ -139,19 +138,19 @@ func UpdateAppNameByRestAPI(serviceAPI string, app models.App, reqLogger logr.Lo
...
 	response, err := client.Do(req)
 	if err != nil || 204 != response.StatusCode {
-		reqLogger.Error(err, "Error when try to do the request", "HTTPMethod", http.MethodPatch, "Request", req, "url", url, "error", err)
+		reqLogger.Error(err, "HTTP StatusCode not expected", "HTTPMethod", http.MethodPatch, "url", url, "response.StatusCode", response.StatusCode)
 		return err
 	}
 	defer response.Body.Close()
 
-	reqLogger.Info("Deleted successfully app object in REST Service API",  "App:", app)
+	reqLogger.Info("Update successfully app name in Rest Service",  "App", app)
```
Comment (laurafitzgerald) [inline pkg/service/client.go]: Suggestiong -> `Successfully updated app name....`

File: pkg/service/client.go [hunk 6]
```diff
@@ -77,44 +78,42 @@ func GetAppFromServiceByRestApi(serviceAPI string, appId string, reqLogger logr.
...
+	if 200 != response.StatusCode && 204 != response.StatusCode {
+		reqLogger.Error(err, "HTTP StatusCode not expected", "HTTPMethod", http.MethodGet, "url", url, "response.StatusCode", response.StatusCode)
 		return got, nil
 	}
 
 	var obj []models.App
 	err = json.NewDecoder(response.Body).Decode(&obj)
 	if err != nil {
-		reqLogger.Error(err, "Error when try to do decoder the body response", "HTTPMethod", http.MethodGet, "Request", req, "Response", response, "url", url, "error", err)
+		reqLogger.Error(err, "Error when try to do decoder the body response", "HTTPMethod", http.MethodGet, "url", url, "Response.Body", response.Body )
```
Comment (laurafitzgerald) [inline pkg/service/client.go]: decoder -> decode

File: pkg/service/client.go [hunk 7]
```diff
@@ -51,20 +51,21 @@ func CreateAppByRestAPI(serviceAPI string, app models.App, reqLogger logr.Logger
...
+	defer response.Body.Close()
+
+	if err != nil || 201 != response.StatusCode {
+		reqLogger.Error(err, "HTTP StatusCode not expected", "HTTPMethod", http.MethodPost, "url", url, "response.StatusCode", response.StatusCode )
 		return err
 	}
-	defer response.Body.Close()
 
-	reqLogger.Info("Created successfully app object in REST Service API",  "App:", app)
+	reqLogger.Info("Created successfully app object in Rest Service",  "App:", app)
```
Comment (laurafitzgerald) [inline pkg/service/client.go]: Suggestion -> `Successfully created app object....`

File: pkg/service/client.go [hunk 8]
```diff
@@ -25,19 +25,19 @@ func DeleteAppFromServiceByRestAPI(serviceAPI string, id string, reqLogger logr.
...
 
-	if err != nil {
-		reqLogger.Error(err, "Error when try to do the request", "HTTPMethod", http.MethodDelete, "Request", req, "url", url, "error", err)
+	if err != nil || 204 != response.StatusCode {
+		reqLogger.Error(err, "HTTP StatusCode not expected", "HTTPMethod", http.MethodDelete, "url", url, "response.StatusCode", response.StatusCode)
 		return err
 	}
 
-	reqLogger.Info("Deleted successfully app object in REST Service API",  "App.Id:", id)
+	reqLogger.Info("Deleted successfully app object from Rest Service",  "App.Id:", id)
```
Comment (laurafitzgerald) [inline pkg/service/client.go]: Suggested -> `Successfully deleted app....`

File: pkg/service/client.go (hunk 3 above)
Comment (camilamacedo86) [reply pkg/service/client.go]: PATCH is the type of request.

File: pkg/service/client.go (hunk 1 above)
Comment (camilamacedo86) [reply pkg/service/client.go]: Done as you suggested.

File: pkg/service/client.go (hunk 2 above)
Comment (camilamacedo86) [reply pkg/service/client.go]: Done

**Comment by laurafitzgerald** [comment]
@camilamacedo86 regarding the verification of this pr, I've been able to deploy the service successfully. Are there other steps that I should take as I don't see the same logs which are shown in the screenshot in the description.

File: pkg/service/client.go (hunk 3 above)
Comment (camilamacedo86) [reply pkg/service/client.go]: 👍 done

File: pkg/service/client.go (hunk 5 above)
Comment (camilamacedo86) [reply pkg/service/client.go]: 👍 done

File: pkg/service/client.go (hunk 6 above)
Comment (camilamacedo86) [reply pkg/service/client.go]: Done 👍

File: pkg/service/client.go (hunk 8 above)
Comment (camilamacedo86) [reply pkg/service/client.go]: 👍 done

**Comment by camilamacedo86** [comment]
HI @laurafitzgerald,

The suggested changes in the typos were made. A new image was built and pushed. 
Note that the logs will appear when the service client be in usage which means when you try to create an app and remove an app in the valid namespace.  I will update the first comment.

File: pkg/service/client.go [hunk 9]
```diff
@@ -77,55 +78,53 @@ func GetAppFromServiceByRestApi(serviceAPI string, appId string, reqLogger logr.
...
 		reqLogger.Info("The app was not found in the REST Service API", "HTTPMethod", http.MethodGet, "url", url)
 		return got, nil
 	}
 
-	reqLogger.Info("App found in the Rest Service", "App", obj[0])
+	reqLogger.Info("App found in the Service", "App", obj[0])
 	return obj[0], nil
 }
 
 //deleteAppFromServiceByRestAPI delete the app object in the service
```
Comment (davidffrench) [inline pkg/service/client.go]: Comment needs to be changed

Comment (camilamacedo86) [reply pkg/service/client.go]: Done. Could we move forward now?

**Comment by camilamacedo86** [comment]
@davidffrench all changes requested are done, could you approve it for we are able to move forward?

**Comment by camilamacedo86** [comment]
Hi @damienomurchu @davidffrench @austincunningham @corinnekrych,

all changes requested are done and/or planned to do in another pr since it not the goal of this one. Could you please check them?

**Pull request merged**